In [13]:
import pandas as pd
import numpy as np
# import matplotlib.pyplot as plt
# import seaborn as sns
from sklearn.preprocessing import OneHotEncoder

[Source](https://wiki.cancerimagingarchive.net/pages/viewpage.action?pageId=70230508#7023050800719a2736ac480481880ca59f36483f)


"...we built a database conducted on 1,775 patients from China with benign or malignant breast disease who underwent mammography examination between July 2012 and January 2016. The database consists of 3,728 mammographies from these 1,775 patients, with biopsy confirmed type of benign or malignant tumors. For 749 of these patients (1,498 mammographies) we also include patients' molecular subtypes. Image data were acquired on a GE Senographe DS mammography system. "

"Clinical data are saved in .XLSX format. Note that for those rows where there exists BOTH a value for ID1 and ID2, TCIA image database stores ONLY the ID2 value as PatientID.
For the D2-XXXX dataset, it is a dataset that only involves malignant tumors. Therefore, only one side of the clinical data is reasonable, such a situation shows that the other side is benign.  We provided mammograms from both the left and right breast."

In [7]:
# df = pd.read_excel('data/CMMD_clinicaldata_revision.xlsx')

In [8]:
# df.to_csv('data/CMMD_clinicaldata_revision.csv', index=False)

In [52]:
df = pd.read_csv('data/CMMD_clinicaldata_revision.csv')
df.head(60)

ID1 LeftRight  Age  number    abnormality classification subtype
0   D1-0001         R   44       2  calcification         Benign     NaN
1   D1-0002         L   40       2  calcification         Benign     NaN
2   D1-0003         L   39       2  calcification         Benign     NaN
3   D1-0004         L   41       2  calcification         Benign     NaN
4   D1-0005         R   42       2  calcification         Benign     NaN
5   D1-0006         L   46       2  calcification         Benign     NaN
6   D1-0007         R   54       2  calcification         Benign     NaN
7   D1-0008         L   38       2  calcification         Benign     NaN
8   D1-0009         L   59       2  calcification         Benign     NaN
9   D1-0010         R   43       2  calcification         Benign     NaN
10  D1-0011         L   27       2  calcification         Benign     NaN
11  D1-0012         L   34       2  calcification         Benign     NaN
12  D1-0013         L   52       2  calcification         Benign     NaN
13  D1-0014         L   52       2  calcification         Benign     NaN
14  D1-0015         R   29       2  calcification         Benign     NaN
15  D1-0016         R   58       2  calcification         Benign     NaN
16  D1-0017         R   45       2  calcification         Benign     NaN
17  D1-0018         R   46       2  calcification         Benign     NaN
18  D1-0019         L   44       2  calcification         Benign     NaN
19  D1-0020         R   53       2  calcification         Benign     NaN
20  D1-0021         L   22       2  calcification         Benign     NaN
21  D1-0021         R   22       2  calcification         Benign     NaN
22  D1-0022         L   34       2  calcification         Benign     NaN
23  D1-0023         R   48       2  calcification         Benign     NaN
24  D1-0024         R   33       2  calcification         Benign     NaN
25  D1-0025         L   49       2  calcification         Benign     NaN
26  D1-0026         R   25       2  calcification         Benign     NaN
27  D1-0027         L   44       2  calcification         Benign     NaN
28  D1-0028         L   35       2  calcification         Benign     NaN
29  D1-0029         L   43       2  calcification         Benign     NaN
30  D1-0030         L   37       2  calcification         Benign     NaN
31  D1-0031         R   51       2  calcification         Benign     NaN
32  D1-0032         L   49       2  calcification         Benign     NaN
33  D1-0032         R   49       2  calcification         Benign     NaN
34  D1-0033         R   57       2  calcification         Benign     NaN
35  D1-0034         L   44       2  calcification         Benign     NaN
36  D1-0034         R   44       2  calcification         Benign     NaN
37  D1-0035         L   45       2  calcification         Benign     NaN
38  D1-0035         R   45       2  calcification         Benign     NaN
39  D1-0036         L   28       2  calcification         Benign     NaN
40  D1-0037         L   41       2  calcification         Benign     NaN
41  D1-0038         R   47       2  calcification         Benign     NaN
42  D1-0039         L   41       2  calcification         Benign     NaN
43  D1-0039         R   41       2  calcification         Benign     NaN
44  D1-0040         L   36       2  calcification         Benign     NaN
45  D1-0041         L   42       2  calcification         Benign     NaN
46  D1-0042         L   41       2  calcification         Benign     NaN
47  D1-0043         L   44       2  calcification         Benign     NaN
48  D1-0045         L   33       2  calcification      Malignant     NaN
49  D1-0046         R   43       2  calcification      Malignant     NaN
50  D1-0047         R   43       2  calcification      Malignant     NaN
51  D1-0051         R   28       2  calcification      Malignant     NaN
52  D1-0053         R   37       2  calcification      Malignant     NaN
53  D1-0055         R   46       2  calcification      Malignant  

In [53]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1872 entries, 0 to 1871
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   ID1             1872 non-null   object
 1   LeftRight       1872 non-null   object
 2   Age             1872 non-null   int64 
 3   number          1872 non-null   int64 
 4   abnormality     1872 non-null   object
 5   classification  1872 non-null   object
 6   subtype         749 non-null    object
dtypes: int64(2), object(5)
memory usage: 102.5+ KB


In [54]:
# Strip leading and trailing spaces from all columns
df = df.apply(lambda x: x.str.strip() if x.dtype == "object" else x)

In [55]:
# Loop over each column and print the unique values
for col in df.columns:
    print(col)
    print(df[col].unique())

ID1
['D1-0001' 'D1-0002' 'D1-0003' ... 'D2-0747' 'D2-0748' 'D2-0749']
LeftRight
['R' 'L']
Age
[44 40 39 41 42 46 54 38 59 43 27 34 52 29 58 45 53 22 48 33 49 25 35 37
 51 57 28 47 36 55 64 56 31 32 63 50 60 62 69 30 26 67 61 68 78 74 70 65
 71 73 66 21 18 17 83 23 24 19 20 84 82 72 80 75 87 85 77 81 76]
number
[2]
abnormality
['calcification' 'both' 'mass']
classification
['Benign' 'Malignant']
subtype
[nan 'Luminal B' 'HER2-enriched' 'Luminal A' 'triple negative']


In [56]:
# Find duplicates in the 'ID1' column
duplicates = df[df.duplicated(subset=['ID1'], keep=False)]

# Count the frequency of each duplicated value in 'ID1'
count = duplicates['ID1'].value_counts()

# Print the duplicated values and their frequency
print("The following values in the 'ID1' column are repeated:")
for value, freq in count.items():
    print(f"{value}: {freq} times")

The following values in the 'ID1' column are repeated:
D1-0021: 2 times
D1-0828: 2 times
D1-0991: 2 times
D1-0990: 2 times
D1-0955: 2 times
D1-0932: 2 times
D1-0931: 2 times
D1-0929: 2 times
D1-0919: 2 times
D1-0908: 2 times
D1-0904: 2 times
D1-0901: 2 times
D1-0894: 2 times
D1-0892: 2 times
D1-0884: 2 times
D1-0876: 2 times
D1-0875: 2 times
D1-0874: 2 times
D1-0869: 2 times
D1-0863: 2 times
D1-0862: 2 times
D1-0853: 2 times
D1-0845: 2 times
D1-0999: 2 times
D1-1009: 2 times
D1-1029: 2 times
D2-0092: 2 times
D2-0701: 2 times
D2-0674: 2 times
D2-0613: 2 times
D2-0565: 2 times
D2-0367: 2 times
D2-0265: 2 times
D2-0259: 2 times
D2-0220: 2 times
D2-0158: 2 times
D2-0090: 2 times
D1-1036: 2 times
D2-0080: 2 times
D2-0072: 2 times
D2-0028: 2 times
D2-0026: 2 times
D1-1673: 2 times
D1-1590: 2 times
D1-1340: 2 times
D1-1220: 2 times
D1-1059: 2 times
D1-0830: 2 times
D1-0820: 2 times
D1-0032: 2 times
D1-0816: 2 times
D1-0274: 2 times
D1-0273: 2 times
D1-0263: 2 times
D1-0257: 2 times
D1-0254: 2

In [57]:
# Find duplicates in the 'ID1' column
duplicates = df[df.duplicated(subset=['ID1'], keep=False)]

# Count the frequency of each duplicated value in 'ID1'
count = duplicates['ID1'].value_counts()

# Get the unique frequency counts
unique_counts = count.unique()

# Print the unique frequency counts
print('The unique frequency counts are:')
print(unique_counts)

The unique frequency counts are:
[2]


In [58]:
# Count the number of duplicate rows
num_duplicates = duplicates.shape[0]

# Print the number of duplicate rows
print(f"There are {num_duplicates} duplicate rows in the 'ID1' column.")

There are 194 duplicate rows in the 'ID1' column.


there are 1775 patients

and 1872 rows

194 duplicate rows / 2 = 97

1872 - 97 = 1775 patients

In [59]:
# Get descriptive statistics for the 'Age' column
age_stats = df['Age'].describe()

# Print the statistics to the console
print('Descriptive statistics for the Age column:')
print(age_stats)

Descriptive statistics for the Age column:
count    1872.000000
mean       47.411859
std        10.873144
min        17.000000
25%        40.000000
50%        46.000000
75%        54.000000
max        87.000000
Name: Age, dtype: float64


In [60]:
# Concatenate the LeftRight column to the end of the ID1 values
df['ID1'] = df['ID1'] + df['LeftRight']

In [61]:
# Convert values in the LeftRight column to 1 or 0
df['LeftRight'] = df['LeftRight'].apply(lambda x: 1 if x == 'R' else 0)

In [62]:
# Rename the LeftRight column to Right
df = df.rename(columns={'LeftRight': 'right'})

In [65]:
df.head()

ID1  right  Age  number    abnormality classification subtype
0  D1-0001R      1   44       2  calcification         Benign     NaN
1  D1-0002L      0   40       2  calcification         Benign     NaN
2  D1-0003L      0   39       2  calcification         Benign     NaN
3  D1-0004L      0   41       2  calcification         Benign     NaN
4  D1-0005R      1   42       2  calcification         Benign     NaN

In [66]:
# Get the unique values in the 'Number' column
unique_values = df['number'].unique()

# Print the unique values to the console
print(unique_values)

[2]


In [67]:
# Drop the 'Number' column
df = df.drop(['number'], axis=1)

In [68]:
# Count the number of instances where 'classification' is 'Malignant' and 'ID1' starts with 'D1'
count = len(df[(df['classification'] == 'Malignant') & df['ID1'].str.startswith('D1')])

# Print the count to the console
print(count)

# Get the total number of ID1 values that start with 'D1'
count = df['ID1'].str.startswith('D1').sum()

# Print the count to the console
print(count)

563
1107


In [69]:
# Count the number of unique values in the ID1 column that start with 'D2'
count = df[df['ID1'].str.startswith('D2')]['ID1'].unique().size

# Print the count to the console
print(count)

765


In [70]:
df.head()

ID1  right  Age    abnormality classification subtype
0  D1-0001R      1   44  calcification         Benign     NaN
1  D1-0002L      0   40  calcification         Benign     NaN
2  D1-0003L      0   39  calcification         Benign     NaN
3  D1-0004L      0   41  calcification         Benign     NaN
4  D1-0005R      1   42  calcification         Benign     NaN

In [72]:
df = df[~df['ID1'].str.startswith('D2')]
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1107 entries, 0 to 1106
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   ID1             1107 non-null   object
 1   right           1107 non-null   int64 
 2   Age             1107 non-null   int64 
 3   abnormality     1107 non-null   object
 4   classification  1107 non-null   object
 5   subtype         0 non-null      object
dtypes: int64(2), object(4)
memory usage: 60.5+ KB


In [74]:
# drop 'subtype' column
df.drop('subtype', axis=1, inplace=True)

# remove 'D1-' from beginning of values in 'ID1' column
df['ID1'] = df['ID1'].str.replace('D1-', '', regex=False).copy()

df.rename(columns={'ID1': 'ID', 'Age': 'age'}, inplace=True)

df.head(60)

KeyError: "['subtype'] not found in axis"

In [75]:
print(df.dtypes)


ID                object
right              int64
age                int64
abnormality       object
classification    object
dtype: object


In [82]:
# df['R'] = df['R'].astype(int)
# df['L'] = df['L'].astype(int)

# make 'abnormality' and 'classification' columns lowercase
df['abnormality'] = df['abnormality'].str.lower()
df['classification'] = df['classification'].str.lower()

# # add '-R' or '-L' suffix to 'ID' based on values in 'R' and 'L' columns
# df.loc[df['R'] == 1, 'ID'] = df['ID'] + '-R'
# df.loc[df['L'] == 1, 'ID'] = df['ID'] + '-L'

df.head()

ID  right  age    abnormality classification
0  0001R      1   44  calcification         benign
1  0002L      0   40  calcification         benign
2  0003L      0   39  calcification         benign
3  0004L      0   41  calcification         benign
4  0005R      1   42  calcification         benign

In [83]:
df.to_csv('data/CMMD_clean.csv', index=False)